### PR TITLE
Update discord.js & fix channel permissions checks

### LINKED ISSUE #2: Text-in-voice support
Currently, TTL does not function on Text-in-voice channels; Likely related to this part of the code: https://github.com/AyuAi/discord-ttl/blob/main/src/app.ts#L106

### PR DESCRIPTION
**Pull request by ayubun**
TTL currently sends a lot of `Missing Access` errors from `GET` requests to the Discord API. This is because I forgot to add a permissions check for `VIEW_CHANNEL` and `READ_MESSAGE_HISTORY`:
Screenshot 2023-04-17 at 3 04 36 PM

Alongside adding these checks, this PR also updates the Discord.JS version (which fixes #2) & adds further permissions checks for unique channel cases (such as TiV, which requires `CONNECT`, on top of `VIEW_CHANNEL` and `READ_MESSAGE_HISTORY`)

### CODE REVIEW

File: src/ttl.ts [hunk 1]
```diff
@@ -3,46 +3,123 @@ import { client } from './app';
...
+      console.error(err);
+    }
+  }
+}
+
+/**
+ * This is a cache so we don't have to keep checking channel permissions all the time.
+ * Channel ID -> Timestamp of when it was added to the dict
+ */
+const channelsWithoutProperPermissions: { [channelId: string]: number } = {};
```
Comment (wasdennnoch) [inline src/ttl.ts:42]: As I mentioned on Discord, there is no need for a custom caching implementation as discord.js itself already caches channels and roles and thus permissions. Methods like `permissionsIn()` further below only access cache anyways. You'd notice if something actually does a network request if a) it's prefixed with `fetch` or b) if it returns a `Promise<>`.

File: src/ttl.ts [hunk 2]
```diff
@@ -3,46 +3,123 @@ import { client } from './app';
...
+  }
+  return true;
+}
+
+function isMessageOlderThanTtl(serverId: string, message: { createdAt: { getTime: () => number } }): boolean {
+  return message.createdAt.getTime() < Date.now() - getServerMessageTtlMillis(serverId);
+}
+
+function canMessageBeBulkDeleted(message: { createdAt: { getTime: () => number } }): boolean {
+  const bulkDeletionThresholdInMillis: number = 1209600000; // 14 days (Discord's bulk deletion age threshold)
```
Comment (wasdennnoch) [inline src/ttl.ts:84]: Maybe to make it easier to read:
```suggestion
  const bulkDeletionThresholdInMillis: number = 1000 * 60 * 60 * 24 * 14; // 14 days (Discord's bulk deletion age threshold)
```

**Review by wasdennnoch — COMMENTED** [review]
The actual date logic checks out, just mostly some code style nits :)

File: src/ttl.ts (hunk 1 above)
Comment (ayubun) [reply src/ttl.ts]: removed

File: src/ttl.ts [hunk 3]
```diff
@@ -3,46 +3,123 @@ import { client } from './app';
...
   while (true) {
     await retrieveMessages();
   }
 }
 
-async function handleDeletesForMessagesOlderThan14Days(
+async function retrieveMessages(): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  for (const item of client.channels.cache) {
+    const channel = item[1];
```
[the remark this answers is not in the document]
Comment (ayubun) [reply src/ttl.ts]: this is such a based tip

**Pull request merged**